### PR TITLE
Delaying ref headers fetch until after startup

### LINKED ISSUE #14615: stats ref codes should not be sent at startup
### Description

ledger.js has code for the refcodes to be sent up right away on startup.
We should be avoiding File IO and network requests at startup time in favour of waiting after some initial delay and/or by lazy loading. 

In this case, this should instead happen when normal checks happen, see 
`runtimeUpdateCheckDelay`

### Steps to Reproduce

  1. Add console.log logging to app/browser/api/ledger.js
  2.Start the browser
  3. Observe output at startup time

**Actual result:**
Network requests happens at startup

**Expected result:**
Network request shouldn't happen directly at startup.

**Reproduces how often:**
Always

### Brave Version
0.23.19 

**about:brave info:**
 
Brave: 0.23.19 
V8: 6.7.288.46 
rev: 178c3fbc045a0cbdbe098db08307503cce952081 
Muon: 7.1.3 
OS Release: 17.5.0 
Update Channel: Release 
OS Architecture: x64 
OS Platform: macOS 
Node.js: 7.9.0 
Tor: 0.3.3.7 (git-035a35178c92da94) 
Brave Sync: v1.4.2 
libchromiumcontent: 67.0.3396.87

**Reproducible on current live release:**
Yes

### PR DESCRIPTION
**Pull request by ryanml**
Fixes #14615 

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).
- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review). (Ask a Brave employee to help if you cannot access this document.)

## Test Plan:

1. Create empty profile folder
2. Create a file called `promoCode` in the profile folder with a test promo code, such as ~`kam253`~
3. Run Brave with `LEDGER_VERBOSE=true and LEDGER_ENVIRONMENT=staging`
4. Confirm promotion code is read after 50 - 70 seconds (you should see call in the terminal)
5. Make sure that `promoCode` file is deleted when using real promo code

## Reviewer Checklist:

- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review) if one was not already requested.

Tests

- [x] Adequate test coverage exists to prevent regressions
- [x] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [ ] New files have MPL2 license header

### CODE REVIEW

File: app/browser/api/ledger.js [hunk 1]
```diff
@@ -2291,6 +2291,16 @@ const fetchReferralHeaders = () => {
   }, {}, appActions.onFetchReferralHeaders)
 }
 
+const scheduleFetchReferralHeaders = () => {
+  const startupFetchDelay = appConfig.ledger.startupFetchDelay
+  // Delay should not occur after client boot has occured
+  const delay = (startupFetchDelay && !bootP) ? startupFetchDelay : 0
+
+  setTimeout(() => {
```
Comment (NejcZdovc) [inline app/browser/api/ledger.js]: We need to track timeout and clear it before we add a new one

Comment (ryanml) [reply app/browser/api/ledger.js]: ++

**Comment by NejcZdovc** [comment]
@bbondy in this PR we are delaying ref request, but we are not delaying initial referral code read (IO). Would you like to delay this as well?

**Comment by ryanml** [comment]
@NejcZdovc in the current flow, delaying the promocode read on startup in turn will delay the call to `fetchReferralHeaders` that was occuring on startup, so I was able to condense the two schedules.

cc: @bbondy

**Comment by NejcZdovc** [comment]
@ryanml please add auto-closing tag

File: app/browser/api/ledger.js [hunk 2]
```diff
@@ -2419,6 +2410,30 @@ const initialize = (state, paymentsEnabled) => {
...
 }
 
+const schedulePromoCodeRead = () => {
+  if (promoCodeReadTimeoutId) {
+    clearTimeout(promoCodeReadTimeoutId)
+  }
+
+  promoCodeReadTimeoutId = setTimeout(() => {
+    module.exports.firstRunPromoCode()
+  }, ledgerUtil.milliseconds.minute)
```
Comment (NejcZdovc) [inline app/browser/api/ledger.js]: this should be random for example 50s to 70s

File: app/browser/api/ledger.js [hunk 3]
```diff
@@ -2343,17 +2344,7 @@ const initialize = (state, paymentsEnabled) => {
...
-      })
-      .catch(error => {
-        if (clientOptions.verboseP) {
-          console.error('read error: ' + error.toString())
-        }
-        fetchReferralHeaders()
-      })
+    schedulePromoCodeRead()
   } else {
     fetchReferralHeaders()
```
Comment (NejcZdovc) [inline app/browser/api/ledger.js]: This should be delayed as well

File: app/browser/api/ledger.js (hunk 2 above)
Comment (ryanml) [reply app/browser/api/ledger.js]: fixed

File: app/browser/api/ledger.js (hunk 3 above)
Comment (ryanml) [reply app/browser/api/ledger.js]: This is now delayed as well

**Comment by ryanml** [comment]
@NejcZdovc changes addressed, ready for review

File: app/browser/api/ledger.js [hunk 4]
```diff
@@ -2419,6 +2407,34 @@ const initialize = (state, paymentsEnabled) => {
   }
 }
 
+const schedulePromoRefFetch = (state) => {
+  if (promoRefFetchTimeoutId) {
+    clearTimeout(promoRefFetchTimeoutId)
+  }
+
+  promoRefFetchTimeoutId = setTimeout(() => {
+    if (updateState.getUpdateProp(state, 'referralDownloadId') == null) {
```
Comment (NejcZdovc) [inline app/browser/api/ledger.js]: state in this case could be already outdated. You have up to 70s old state here

**Comment by NejcZdovc** [comment]
@ryanml please add test plan

File: app/browser/api/ledger.js (hunk 4 above)
Comment (ryanml) [reply app/browser/api/ledger.js]: App action has been created to remedy this, a fresh state will be used in these actions now.

**Comment by jasonrsadler** [comment]
@ryanml Can you rebase/update?

**Comment by ryanml** [comment]
@jasonrsadler rebased

File: app/browser/api/ledger.js [hunk 5]
```diff
@@ -2347,21 +2348,8 @@ const initialize = (state, paymentsEnabled) => {
...
-        if (clientOptions.verboseP) {
-          console.error('read error: ' + error.toString())
-        }
-        fetchReferralHeaders()
-      })
-  } else {
-    fetchReferralHeaders()
-  }
+  // Schedule promo run and referral header fetch
+  schedulePromoRefFetch(state)
```
Comment (jasonrsadler) [inline app/browser/api/ledger.js]: I think we should move this call to the APP_INITIALIZED hook. in here:
https://github.com/brave/browser-laptop/blob/master/app/browser/reducers/ledgerReducer.js#L613
This gets called after most of the browser startup work and ledger.initialize and could save us a little bit on startup :) wdyt?

Comment (ryanml) [reply app/browser/api/ledger.js]: I like it!

Comment (ryanml) [reply app/browser/api/ledger.js]: @jasonrsadler done. Also I can confirm that I am receiving 400 for the promo PUT using the test code on master.

**Comment by ryanml** [comment]
@jasonrsadler am receiving a 400 on promo PUT as well on master, still see what's going on with things

@NejcZdovc

**Comment by ryanml** [comment]
Further inspecting the 400 that occurs on PUT to `/promo/initialize/nonua`, I am seeing that `referralApi` (even when running non-staging) is always an empty string. The response from the endpoint is:

`"child \"api_key\" fails because [\"api_key\" is required]"`

The referralApi key should either be extracted from the buildConfig (you would need to package) or provided via `LEDGER_REFERRAL_API_KEY` so this is probably why we are running in to issues.

**Comment by NejcZdovc** [comment]
updated test plan to the real promoCode key

**Comment by NejcZdovc** [comment]
master 2483444
0.24 ed46f9874
0.23 bc030f67a

**Pull request merged**